db/migrate/20180813125658_create_languages.rb

```diff
@@ -0,0 +1,13 @@
+class CreateLanguages < ActiveRecord::Migration[5.1]
+  def change
+    create_table :languages do |t|
+      t.string :name
+      t.timestamp
+    end
+
+    create_table :languages_projects, id: false do |t|
+      t.belongs_to :project, index: true
+      t.belongs_to :language, index: true, unique: true
```

Comment:
I don't think the `unique: true` will have the desired effect here.

I think that we probably want the following:

```rb
class CreateLanguages < ActiveRecord::Migration[5.1]
  def change
    create_table :languages do |t|
      t.string :name, unique: true # <-- to ensure that language names are unique
      t.timestamp
    end
    create_table :languages_projects, id: false do |t|
      t.belongs_to :project, index: true
      t.belongs_to :language, index: true
    end
    add_index :languages_projects, [:project_id, :language_id], :unique => true # <-- to ensure that language project relations are unique
  end
end
```

Reply:
Ok, I have changed it like above. The validations in the models will only rais an error if someone tries to create a language that is not unique, but these validations at the database level will ensure no one is able to append any languages which are not unique, is that a correct understanding of the difference/ need for both? Or should there be refactoring of the model validations?